### PR TITLE
Added the Commit phase

### LINKED ISSUE #145: Add a Commit phase to the Interceptor interface
As some plugins might want to modify/add headers or modify the body (e.g. CSP/XSRF nonce injection in HTML templates) exactly before the response is written, the interceptor interface needs to support a Commit phase.

### PR DESCRIPTION
**Pull request by mattiasgrenfeldt**
Fixes #145

Added the `Commit` phase to the Interceptor interface. This will be called by the `safehttp.ResponseWriter` before a response is written to the underlying `http.ResponseWriter`, for all interceptors that were installed on the handler. If a response  is written in any of the `Commit` phases,  then the `Commit` phases of the remaining interceptor won't execute.

As the `Commit` phase takes a `safehttp.Response` as argument (e.g. for injecting nonces), we also created a `safehttp.TemplateResponse` interface that encapsulates a safe template and the data that is applied to it.

Co-authored-by: Mihali Mara <maramihali@google.com>

### CODE REVIEW

**Comment by kele** [comment]
What's the need for `CommitResponseWriter`? Why can't it be just `ResponseWriter`? If you plan to change something about the `ResponseWriter`, can it be done in a separate PR?

Similarly, can the `DataTemplate` be done in a separate PR?

**Comment by mihalimara22** [comment]
We want the `Commit` phase to only have access to the `WriteError` method in the `ResponseWriter` and the ability to set headers. We considered some alternatives, but they were either very messy or required big changes to the whole design. So, we created the `CommitResponseWriter`, which is basically a wrapper around `ResponseWriter` that serves the purpose above. I think this should be in this PR since it doesn't make much sense without the `Commit` phase (which is its only use case). Similarly, the `DataTemplate` is a `Response` that bundles together `Template` and `data` so it can be passed to the `Commit` phase which might want to modify them (e.g. inject nonces). So it is also strictly related to the `Commit` phase (for now at least). 

They do need documentation, but we wanted to do that together with the documentation for `ResponseWriter`, `IncomingRequest`, etc.

 WDYT? Should they still be split?

**Comment by kele** [comment]
> We want the `Commit` phase to only have access to the `WriteError` method in the `ResponseWriter` and the ability to set headers. (...)

Please introduce the `Commit` phase without any new features first.

> Similarly, the `DataTemplate` is a `Response` that bundles together `Template` and `data` so it can be passed to the `Commit` phase which might want to modify them (e.g. inject nonces). So it is also strictly related to the `Commit` phase (for now at least).

Sure, this seems unavoidable. I didn't think of it, sorry.

> WDYT? Should they still be split?

Yes, please just keep a simple `Commit` phase with support for whatever features we have right now (i.e. that's why `DataTemplate` is okay).

**Comment by maramihali** [comment]
So the alternative would be to have the `ResponseWriter` passed as argument to the `Commit` phase and then in a subsequent PR have the `CommitResponseWriter` (or something with a better name). Would you prefer it that way (just to make sure I understood it correctly)?

**Comment by kele** [comment]
Yup! From my perspective:

- `Commit` phase needs to be added
- `CommitResponseWriter` or whatever other restrictions regarding what `Commit` phase can and cannot do are a separate feature
- `DataTemplate` seems to be required, as adding `Commit` with support only for `Write` and not `ExecuteTemplate` seems like introducing an inconsistency
- I don't understand what `CommitPhase` does

File: safehttp/mux.go [hunk 1]
```diff
@@ -198,8 +198,14 @@ type handlerWithInterceptors struct {
 // ServeHTTP calls the Before method of all the interceptors and then calls the
 // underlying handler.
```
Comment (empijei) [inline safehttp/mux.go:199]: Change docs accordingly.

File: safehttp/mux_test.go [hunk 2]
```diff
@@ -324,16 +400,23 @@ func (setHeaderConfig) Match(i safehttp.Interceptor) bool {
...
-	w.Header().Set(name, value)
+	w.Header().Set("Pizza", value)
 	return safehttp.Result{}
 }
 
+func (p setHeaderConfigInterceptor) Commit(w *safehttp.CommitResponseWriter, r *safehttp.IncomingRequest, resp safehttp.Response, cfg interface{}) safehttp.Result {
+	value := "Fusili"
+	if c, ok := cfg.(setHeaderConfig); ok {
+		value = c.pastaValue
+	}
```
Comment (empijei) [inline safehttp/mux_test.go:415]: I would honestly use a type assertion without check. If this assertion fails we have a big bug in the framework and I'd rather have this panic.

**Review by empijei — COMMENTED** [review]
I second kele comments, and I would add that we will also need a wrapper for JSON, not just templates, but I would suggest to call both JSONResponse and TemplateResponse. DataTemplate is not clear to me as a name.

I also think that the CommitPhase type might be superfluous, can't we do this by adding to types we already have?

File: safehttp/mux.go [hunk 3]
```diff
@@ -183,15 +184,17 @@ func (m methodHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
...
 	handler   Handler
-	interceps []appliedInterceptor
+	interceps []AppliedInterceptor
 	disp      Dispatcher
 }
 
 // ServeHTTP calls the Before method of all the interceptors and then calls the
-// underlying handler.
+// underlying handler. Any panics that occur during the Before phase, during handler
+// execution or during the Commit phase are recovered here and a 500 Internal Server
```
Comment (stapelberg) [inline safehttp/mux.go:193]: Recovering panics is a design decision that the writers of `net/http` regretted.

Wouldn’t it be safer to not make that mistake again?

Even if the answer is no, the comment should be updated to explain.

File: safehttp/response_writer.go [hunk 4]
```diff
@@ -53,21 +57,54 @@ func NotWritten() Result {
 	return Result{}
 }
 
+// commitPhase calls the Commit phases of all the interceptors. This stage will
+// run before a response is written to the ResponseWriter. If a response is
+// written to the ResponseWriter in a Commit phase then the Commit phases of the
+// remaining interceptors won't execute.
+//
+// TODO: BIG WARNING, if ResponseWriter.Write and ResponseWriter.WriteTemplate
+// are called in Commit then this will recurse. CommitResponseWriter was an
```
Comment (stapelberg) [inline safehttp/response_writer.go:66]: What happened to CommitResponseWriter? Don’t leave the reader hanging :)

File: safehttp/mux.go (hunk 3 above)
Comment (maramihali) [reply safehttp/mux.go:193]: Finding a better alternative to panics is one of our TODOs so they will disappear in v0 :)

File: safehttp/response_writer.go (hunk 4 above)
Comment (maramihali) [reply safehttp/response_writer.go:66]: @kele suggested to leave this feature (and other possible restrictions to the `Commit` phase that we haven't foreseen) for another PR. :)

File: safehttp/mux_test.go [hunk 5]
```diff
@@ -315,8 +371,8 @@ func TestMuxInterceptors(t *testing.T) {
 }
 
 type setHeaderConfig struct {
-	name  string
-	value string
```
Comment (kele) [inline safehttp/mux_test.go:319]: Please remove all unrelated changes.

File: safehttp/mux.go [hunk 6]
```diff
@@ -96,7 +96,8 @@ func NewServeMux(d Dispatcher, domains ...string) *ServeMux {
 	}
 }
 
-type appliedInterceptor struct {
+// AppliedInterceptor TODO
+type AppliedInterceptor struct {
```
Comment (kele) [inline safehttp/mux.go:100]: Why is this part of the public API now?

File: safehttp/response.go [hunk 7]
```diff
@@ -23,3 +23,12 @@ type Response interface{}
 type Template interface {
 	Execute(wr io.Writer, data interface{}) error
 }
+
+// TemplateResponse TODO
```
Comment (kele) [inline safehttp/response.go:27]: Fix the TODOs.

File: safehttp/response_writer.go [hunk 8]
```diff
@@ -25,18 +25,22 @@ type ResponseWriter struct {
 
 	// Having this field unexported is essential for security. Otherwise one can
 	// easily overwrite the struct bypassing all our safety guarantees.
-	header  Header
-	written bool
+	header           Header
+	appliedInterceps []AppliedInterceptor
+	req              *IncomingRequest
```
Comment (kele) [inline safehttp/response_writer.go:30]: Why is this a part of a `ResponseWriter` now?

File: safehttp/response_writer.go [hunk 9]
```diff
@@ -25,18 +25,22 @@ type ResponseWriter struct {
...
+	written          bool
 }
 
 // NewResponseWriter creates a ResponseWriter from a safehttp.Dispatcher, an
 // http.ResponseWriter and a list of interceptors associated with a ServeMux.
-func NewResponseWriter(d Dispatcher, rw http.ResponseWriter) *ResponseWriter {
+func NewResponseWriter(d Dispatcher, rw http.ResponseWriter, req *IncomingRequest, interceps []AppliedInterceptor) *ResponseWriter {
```
Comment (kele) [inline safehttp/response_writer.go:36]: It doesn't seem right that the `ResponseWriter` has direct control over the interceptors. It should ask the mux or the handler wrapper that has the interceptors to do the work.

File: safehttp/response_writer.go [hunk 10]
```diff
@@ -73,13 +110,20 @@ func (w *ResponseWriter) WriteTemplate(t Template, data interface{}) Result {
...
+		return Result{}
+	}
 	w.markWritten()
 	w.rw.WriteHeader(int(StatusNoContent))
 	return Result{}
 }
 
 // WriteError writes an error response (400-599) according to the provided status
-// code.
+// code. Any headers previously set will be removed.
```
Comment (kele) [inline safehttp/response_writer.go:126]: Is this still true?

File: safehttp/response_writer.go (hunk 8 above)
Comment (mattiasgrenfeldt) [reply safehttp/response_writer.go:30]: Well, both `appliedInterceps` and `req` needed to be part of the `ResponseWriter` for our design where the `ResponseWriter` would run the `commitPhase`. In the newer design I'm working on now the `ResponseWriter` still needs to own the `req` but instead of `appliedInterceps` it has a reference to the `handlerWithInterceptors` and the `commitPhase` is moved to the `handlerWithInterceptors`.

File: safehttp/mux.go (hunk 6 above)
Comment (mattiasgrenfeldt) [reply safehttp/mux.go:100]: It was part of the public API since we wanted to pass it as part of the `ResponseWriter` constructor and for that it needs to be exported, so that the constructor doesn't become unusable outside the package.

File: safehttp/mux_test.go (hunk 5 above)
Comment (mattiasgrenfeldt) [reply safehttp/mux_test.go:319]: What do you mean with unrelated changes? We have updated the `SetHeaderInterceptor with config` and the `SetHeaderInterceptor with mismatching config` tests to also test the commit phase. Earlier they only tested the Before phase. How is this unrelated from the PR introducing the commit phase?

I looked through all our changes and found one unrelated change in `mux_test.go` in `TestMuxInterceptors`:
```
- wantStatus: 500,
+ wantStatus: safehttp.StatusInternalServerError,
```
We have in an earlier PR replaced all numeric status codes with constants and apparently I had missed this instance. How could it not be ok to fix these kinds of tiny nits when they are discovered? The other option would be to open a new issue and PR to just change this little thing.

File: safehttp/mux_test.go (hunk 2 above)
Comment (mattiasgrenfeldt) [reply safehttp/mux_test.go]: Ah, yes, this check was to check if a valid config was passed. If there was no valid config passed then the default values would be used.

File: safehttp/mux_test.go (hunk 5 above)
Comment (mattiasgrenfeldt) [reply safehttp/mux_test.go]: Fixed!

File: safehttp/response_writer.go [hunk 11]
```diff
@@ -26,17 +26,22 @@ type ResponseWriter struct {
...
 	written bool
 }
 
 // NewResponseWriter creates a ResponseWriter from a safehttp.Dispatcher, an
-// http.ResponseWriter and a list of interceptors associated with a ServeMux.
-func NewResponseWriter(d Dispatcher, rw http.ResponseWriter) *ResponseWriter {
+// http.ResponseWriter and a reference to the current IncomingRequest being served.
+// The IncomingRequest will only be used by the commit phase, which only runs when
+// the ServeMux is used, and can be passed as nil in tests.
+func NewResponseWriter(d Dispatcher, rw http.ResponseWriter, req *IncomingRequest) *ResponseWriter {
```
Comment (kele) [inline safehttp/response_writer.go:38]: Let's finish this PR, but the `IncomingRequest` should not be added here, please leave a TODO.

**Review by kele — APPROVED** [review]

File: safehttp/response_writer.go (hunk 11 above)
Comment (kele) [inline safehttp/response_writer.go:38]:
```suggestion
// TODO: remove the IncomingRequest parameter.
func NewResponseWriter(d Dispatcher, rw http.ResponseWriter, req *IncomingRequest) *ResponseWriter {
```
